cmd/consumer/main.go

```diff
@@ -0,0 +1,230 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/alexdunne/gs-onboarding/hn"
+	"github.com/alexdunne/gs-onboarding/internal/postgres"
+	"github.com/pkg/errors"
+	"github.com/spf13/viper"
+	"go.uber.org/zap"
+)
+
+type Config struct {
+	WorkerCount            int
+	WorkerIntervalDuration time.Duration
+	DatabaseDSN            string
+}
+
+func loadConfig() (*Config, error) {
+	viper.SetConfigFile(".env")
+	if err := viper.ReadInConfig(); err != nil {
+		return nil, errors.Wrap(err, "failed to read env file")
+	}
+
+	c := &Config{
+		WorkerCount:            runtime.NumCPU(),
+		WorkerIntervalDuration: 300 * time.Second,
+		DatabaseDSN: fmt.Sprintf(
+			"postgres://%s:%s@%s:%s/%s",
+			viper.GetString("DATABASE_USER"),
+			viper.GetString("DATABASE_PASSWORD"),
+			viper.GetString("DATABASE_HOST"),
+			viper.GetString("DATABASE_PORT"),
+			viper.GetString("DATABASE_DB"),
+		),
+	}
+
+	intervalSeconds := viper.GetInt("WORKER_INTERVAL_SECONDS")
+	if intervalSeconds != 0 {
+		c.WorkerIntervalDuration = time.Duration(intervalSeconds) * time.Second
+	}
+
+	return c, nil
+}
+
+func main() {
+	if err := run(); err != nil {
+		log.Fatal(err)
+		os.Exit(1)
+	}
+}
+
+func run() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	cfg, err := loadConfig()
+	if err != nil {
+		return errors.Wrap(err, "loading config")
+	}
+
+	logger, err := zap.NewProduction()
+	if err != nil {
+		return errors.Wrap(err, "creating logger")
+	}
+	defer logger.Sync()
+
+	go func() {
+		// run the consumer the first time when the goroutine starts
+		consumer(ctx, cfg, logger)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.Tick(cfg.WorkerIntervalDuration):
+				logger.Info("running consumer")
+				consumer(ctx, cfg, logger)
+			}
+		}
+	}()
+
+	// handle interrupts and propagate the changes across the consumer pipeline
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	<-c
+	cancel()
```

Comment:
You can actually pull this into a goroutine and take your for loop out, this avoids you having to nest all of your logic in go routines, and just run it on the main thread.

i.e.

```
func main() {
	ctx, cancelFn := context.WithCancel(context.Background())
        defer cancelFn()
        
	go func() {
		c := make(chan os.Signal, 1)
		signal.Notify(c, os.Interrupt)
		<-c
		cancelFn()
	}()
	
	for {
		select {
		case <-ctx.Done():
			return
		case <-time.Tick(cfg.WorkerIntervalDuration):
			logger.Info("running consumer")
			consumer(ctx, cfg, logger)
		}
	}
}
```

Reply:
Nice 👍 I tried doing that originally but completely blanked on moving the `for` out of the goroutine so it was cancelling immediately.